File: API/1918_Trend_Collector/CS/TrendCollectorStrategy.cs
using System;
using System.Collections.Generic;

using Ecng.Common;

using StockSharp.Algo.Indicators;
using StockSharp.Algo.Strategies;
using StockSharp.BusinessEntities;
using StockSharp.Messages;

namespace StockSharp.Samples.Strategies;

/// <summary>
/// Trend following strategy with stochastic and EMA crossover.
/// Buys when fast EMA above slow EMA and stochastic oversold.
/// Sells when fast EMA below slow EMA and stochastic overbought.
/// </summary>
public class TrendCollectorStrategy : Strategy
{
	private readonly StrategyParam<int> _fastMaLength;
	private readonly StrategyParam<int> _slowMaLength;
	private readonly StrategyParam<decimal> _stochasticUpper;
	private readonly StrategyParam<decimal> _stochasticLower;
	private readonly StrategyParam<DataType> _candleType;

	private ExponentialMovingAverage _fastMa;
	private ExponentialMovingAverage _slowMa;
	private StochasticOscillator _stochastic;

	public int FastMaLength { get => _fastMaLength.Value; set => _fastMaLength.Value = value; }
	public int SlowMaLength { get => _slowMaLength.Value; set => _slowMaLength.Value = value; }
	public decimal StochasticUpper { get => _stochasticUpper.Value; set => _stochasticUpper.Value = value; }
	public decimal StochasticLower { get => _stochasticLower.Value; set => _stochasticLower.Value = value; }
	public DataType CandleType { get => _candleType.Value; set => _candleType.Value = value; }

	public TrendCollectorStrategy()
	{
		_fastMaLength = Param(nameof(FastMaLength), 4)
			.SetGreaterThanZero()
			.SetDisplay("Fast EMA Length", "Fast EMA length", "Parameters");

		_slowMaLength = Param(nameof(SlowMaLength), 50)
			.SetGreaterThanZero()
			.SetDisplay("Slow EMA Length", "Slow EMA length", "Parameters");

		_stochasticUpper = Param(nameof(StochasticUpper), 60m)
			.SetDisplay("Stochastic Upper", "Upper stochastic level", "Parameters");

		_stochasticLower = Param(nameof(StochasticLower), 40m)
			.SetDisplay("Stochastic Lower", "Lower stochastic level", "Parameters");

		_candleType = Param(nameof(CandleType), TimeSpan.FromHours(1).TimeFrame())
			.SetDisplay("Candle Type", "Type of candles", "General");
	}

	/// <inheritdoc />
	public override IEnumerable<(Security sec, DataType dt)> GetWorkingSecurities()
	{
		return [(Security, CandleType)];
	}

	/// <inheritdoc />
	protected override void OnReseted()
	{
		base.OnReseted();
		_fastMa = default;
		_slowMa = default;
		_stochastic = default;
	}

	/// <inheritdoc />
	protected override void OnStarted2(DateTime time)
	{
		base.OnStarted2(time);

		_fastMa = new ExponentialMovingAverage { Length = FastMaLength };
		_slowMa = new ExponentialMovingAverage { Length = SlowMaLength };
		_stochastic = new StochasticOscillator
		{
			K = { Length = 14 },
			D = { Length = 3 },
		};

		var subscription = SubscribeCandles(CandleType);
		subscription
			.BindEx(_stochastic, ProcessCandle)
			.Start();

		var area = CreateChartArea();
		if (area != null)
		{
			DrawCandles(area, subscription);
			DrawOwnTrades(area);
		}
	}

	private void ProcessCandle(ICandleMessage candle, IIndicatorValue stochValue)
	{
		if (candle.State != CandleStates.Finished)
			return;

		var fastResult = _fastMa.Process(candle.ClosePrice, candle.OpenTime, true);
		var slowResult = _slowMa.Process(candle.ClosePrice, candle.OpenTime, true);

		if (!fastResult.IsFormed || !slowResult.IsFormed || !stochValue.IsFormed)
			return;

		var fast = fastResult.ToDecimal();
		var slow = slowResult.ToDecimal();

		var stoch = (StochasticOscillatorValue)stochValue;
		if (stoch.K is not decimal stochK)
			return;

		// Buy: fast EMA above slow EMA, stochastic oversold
		if (Position <= 0 && fast > slow && stochK < StochasticLower)
		{
			if (Position < 0) BuyMarket();
			BuyMarket();
		}
		// Sell: fast EMA below slow EMA, stochastic overbought
		else if (Position >= 0 && fast < slow && stochK > StochasticUpper)
		{
			if (Position > 0) SellMarket();
			SellMarket();
		}
	}
}

File: API/1918_Trend_Collector/PY/trend_collector_strategy.py
import clr

clr.AddReference("StockSharp.Messages")
clr.AddReference("StockSharp.Algo")

from System import TimeSpan
from StockSharp.Messages import DataType, CandleStates
from StockSharp.Algo.Indicators import ExponentialMovingAverage, StochasticOscillator
from StockSharp.Algo.Strategies import Strategy


class trend_collector_strategy(Strategy):
    def __init__(self):
        super(trend_collector_strategy, self).__init__()
        self._fast_ma_length = self.Param("FastMaLength", 4) \
            .SetGreaterThanZero() \
            .SetDisplay("Fast EMA Length", "Fast EMA length", "Parameters")
        self._slow_ma_length = self.Param("SlowMaLength", 50) \
            .SetGreaterThanZero() \
            .SetDisplay("Slow EMA Length", "Slow EMA length", "Parameters")
        self._stochastic_upper = self.Param("StochasticUpper", 60.0) \
            .SetDisplay("Stochastic Upper", "Upper stochastic level", "Parameters")
        self._stochastic_lower = self.Param("StochasticLower", 40.0) \
            .SetDisplay("Stochastic Lower", "Lower stochastic level", "Parameters")
        self._candle_type = self.Param("CandleType", DataType.TimeFrame(TimeSpan.FromHours(1))) \
            .SetDisplay("Candle Type", "Type of candles", "General")
        self._fast_ma = None
        self._slow_ma = None

    @property
    def fast_ma_length(self):
        return self._fast_ma_length.Value
    @property
    def slow_ma_length(self):
        return self._slow_ma_length.Value
    @property
    def stochastic_upper(self):
        return self._stochastic_upper.Value
    @property
    def stochastic_lower(self):
        return self._stochastic_lower.Value
    @property
    def candle_type(self):
        return self._candle_type.Value

    def OnReseted(self):
        super(trend_collector_strategy, self).OnReseted()

    def OnStarted(self, time):
        super(trend_collector_strategy, self).OnStarted(time)
        self._fast_ma = ExponentialMovingAverage()
        self._fast_ma.Length = self.fast_ma_length
        self._slow_ma = ExponentialMovingAverage()
        self._slow_ma.Length = self.slow_ma_length
        stochastic = StochasticOscillator()
        stochastic.K.Length = 14
        stochastic.D.Length = 3
        subscription = self.SubscribeCandles(self.candle_type)
        subscription.BindEx(stochastic, self.process_candle).Start()
        area = self.CreateChartArea()
        if area is not None:
            self.DrawCandles(area, subscription)
            self.DrawOwnTrades(area)

    def process_candle(self, candle, stoch_value):
        if candle.State != CandleStates.Finished:
            return
        fast_result = self._fast_ma.Process(candle.ClosePrice, candle.OpenTime, True)
        slow_result = self._slow_ma.Process(candle.ClosePrice, candle.OpenTime, True)
        if not fast_result.IsFormed or not slow_result.IsFormed or not stoch_value.IsFormed:
            return
        fast = float(fast_result)
        slow = float(slow_result)
        k_val = stoch_value.K
        if k_val is None:
            return
        stoch_k = float(k_val)
        # Buy: fast EMA above slow EMA, stochastic oversold
        if self.Position <= 0 and fast > slow and stoch_k < float(self.stochastic_lower):
            if self.Position < 0:
                self.BuyMarket()
            self.BuyMarket()
        # Sell: fast EMA below slow EMA, stochastic overbought
        elif self.Position >= 0 and fast < slow and stoch_k > float(self.stochastic_upper):
            if self.Position > 0:
                self.SellMarket()
            self.SellMarket()

    def CreateClone(self):
        return trend_collector_strategy()
